File: benchmark/languages/c/work/day06/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_WIDTH 200000

int main(int argc, char *argv[]) {
    int do_part1 = 1, do_part2 = 1;
    if (argc >= 2) {
        if (argv[1][0] == '1') do_part2 = 0;
        else if (argv[1][0] == '2') do_part1 = 0;
    }

    FILE *f = fopen("input.txt", "r");
    if (!f) { fprintf(stderr, "Cannot open input.txt\n"); return 1; }

    // Read all lines
    char **lines = NULL;
    int nlines = 0;
    int capacity = 0;
    char *buf = malloc(MAX_WIDTH + 2);

    while (fgets(buf, MAX_WIDTH + 1, f)) {
        // Remove trailing newline/CR
        int len = (int)strlen(buf);
        while (len > 0 && (buf[len-1] == '\n' || buf[len-1] == '\r')) {
            buf[--len] = '\0';
        }
        if (nlines >= capacity) {
            capacity = capacity ? capacity * 2 : 8;
            lines = realloc(lines, capacity * sizeof(char*));
        }
        lines[nlines] = malloc(len + 1);
        memcpy(lines[nlines], buf, len + 1);
        nlines++;
    }
    free(buf);
    fclose(f);

    // Strip trailing empty lines
    while (nlines > 0 && lines[nlines-1][0] == '\0') {
        free(lines[nlines-1]);
        nlines--;
    }

    if (nlines == 0) {
        if (do_part1) printf("Part 1: 0\n");
        if (do_part2) printf("Part 2: 0\n");
        free(lines);
        return 0;
    }

    // Find max width
    int width = 0;
    for (int i = 0; i < nlines; i++) {
        int len = (int)strlen(lines[i]);
        if (len > width) width = len;
    }

    // The last row is the operator row
    int op_row = nlines - 1;

    // Identify separator columns: columns where ALL rows have ' ' or are out of bounds
    int *is_sep = calloc(width, sizeof(int));
    for (int col = 0; col < width; col++) {
        int sep = 1;
        for (int row = 0; row < nlines; row++) {
            int rowlen = (int)strlen(lines[row]);
            char c = (col < rowlen) ? lines[row][col] : ' ';
            if (c != ' ') { sep = 0; break; }
        }
        is_sep[col] = sep;
    }

    // Find problem groups: ranges [start, end) of consecutive non-separator columns
    typedef struct { int start; int end; } Problem;
    Problem *probs = NULL;
    int nprobs = 0;
    int pcap = 0;

    int in_prob = 0;
    int prob_start = 0;
    for (int col = 0; col <= width; col++) {
        int sep = (col == width) ? 1 : is_sep[col];
        if (!sep && !in_prob) {
            in_prob = 1;
            prob_start = col;
        } else if (sep && in_prob) {
            in_prob = 0;
            if (nprobs >= pcap) {
                pcap = pcap ? pcap * 2 : 64;
                probs = realloc(probs, pcap * sizeof(Problem));
            }
            probs[nprobs].start = prob_start;
            probs[nprobs].end = col;
            nprobs++;
        }
    }

    // --- Part 1 ---
    // Within each problem: each data row (0..op_row-1) gives one number
    // by concatenating non-space chars across all columns of the problem.
    // Operator is in op_row (any column of the problem that has '*' or '+').
    long long grand_total_1 = 0;
    if (do_part1) {
        for (int p = 0; p < nprobs; p++) {
            int s = probs[p].start;
            int e = probs[p].end;

            // Find operator
            char op = ' ';
            for (int col = s; col < e; col++) {
                int rowlen = (int)strlen(lines[op_row]);
                char c = (col < rowlen) ? lines[op_row][col] : ' ';
                if (c == '*' || c == '+') { op = c; break; }
            }
            if (op == ' ') continue;

            // For each data row, collect digits across columns s..e-1
            long long result = 0;
            int first = 1;
            for (int row = 0; row < op_row; row++) {
                int rowlen = (int)strlen(lines[row]);
                char numbuf[64];
                int nlen = 0;
                for (int col = s; col < e && nlen < 63; col++) {
                    char c = (col < rowlen) ? lines[row][col] : ' ';
                    if (c >= '0' && c <= '9') {
                        numbuf[nlen++] = c;
                    }
                }
                if (nlen == 0) continue;
                numbuf[nlen] = '\0';
                long long val = atoll(numbuf);
                if (first) {
                    result = val;
                    first = 0;
                } else {
                    if (op == '*') result *= val;
                    else result += val;
                }
            }
            grand_total_1 += result;
        }
    }

    // --- Part 2 ---
    // Within each problem: each column gives one number
    // by concatenating non-space chars from top to bottom (data rows only).
    // Operator is in the leftmost column of the problem (op_row).
    // Problems are read right-to-left (but operator is in leftmost col).
    // Within a problem, columns are read right-to-left.
    long long grand_total_2 = 0;
    if (do_part2) {
        // Process problems in any order (grand total is sum), but columns within problem right-to-left
        for (int p = 0; p < nprobs; p++) {
            int s = probs[p].start;
            int e = probs[p].end;

            // Find operator: in op_row, leftmost column of the problem
            char op = ' ';
            for (int col = s; col < e; col++) {
                int rowlen = (int)strlen(lines[op_row]);
                char c = (col < rowlen) ? lines[op_row][col] : ' ';
                if (c == '*' || c == '+') { op = c; break; }
            }
            if (op == ' ') continue;

            // Read columns from right (e-1) to left (s), each column is a number
            long long result = 0;
            int first = 1;
            for (int col = e - 1; col >= s; col--) {
                char numbuf[64];
                int nlen = 0;
                for (int row = 0; row < op_row && nlen < 63; row++) {
                    int rowlen = (int)strlen(lines[row]);
                    char c = (col < rowlen) ? lines[row][col] : ' ';
                    if (c >= '0' && c <= '9') {
                        numbuf[nlen++] = c;
                    }
                }
                if (nlen == 0) continue;
                numbuf[nlen] = '\0';
                long long val = atoll(numbuf);
                if (first) {
                    result = val;
                    first = 0;
                } else {
                    if (op == '*') result *= val;
                    else result += val;
                }
            }
            grand_total_2 += result;
        }
    }

    if (do_part1) printf("Part 1: %lld\n", grand_total_1);
    if (do_part2) printf("Part 2: %lld\n", grand_total_2);

    // Cleanup
    free(is_sep);
    free(probs);
    for (int i = 0; i < nlines; i++) free(lines[i]);
    free(lines);

    return 0;
}

File: benchmark/languages/go/work/day06/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readLines(filename string) []string {
	f, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	// Remove trailing empty lines
	for len(lines) > 0 && strings.TrimSpace(lines[len(lines)-1]) == "" {
		lines = lines[:len(lines)-1]
	}
	return lines
}

// padLines pads all lines to the same length
func padLines(lines []string) []string {
	maxLen := 0
	for _, l := range lines {
		if len(l) > maxLen {
			maxLen = len(l)
		}
	}
	result := make([]string, len(lines))
	for i, l := range lines {
		if len(l) < maxLen {
			result[i] = l + strings.Repeat(" ", maxLen-len(l))
		} else {
			result[i] = l
		}
	}
	return result
}

// isAllSpaces checks if all lines have a space at column col
func isAllSpaces(lines []string, col int) bool {
	for _, l := range lines {
		if col < len(l) && l[col] != ' ' {
			return false
		}
	}
	return true
}

// parseProblems splits the grid into problems.
// Each problem is a slice of column indices (contiguous non-all-space columns).
func parseProblems(lines []string) [][]int {
	if len(lines) == 0 {
		return nil
	}
	maxLen := 0
	for _, l := range lines {
		if len(l) > maxLen {
			maxLen = len(l)
		}
	}

	var problems [][]int
	var current []int
	for col := 0; col < maxLen; col++ {
		if isAllSpaces(lines, col) {
			if len(current) > 0 {
				problems = append(problems, current)
				current = nil
			}
		} else {
			current = append(current, col)
		}
	}
	if len(current) > 0 {
		problems = append(problems, current)
	}
	return problems
}

// Part 1: Each problem's columns together form numbers (read normally as horizontal text).
// The last row is the operator. Numbers are in rows 0..numRows-2.
// Extract the text for each number row by taking the columns of that problem,
// then parse the trimmed string as a number.
func solvePart1(lines []string) int64 {
	lines = padLines(lines)
	problems := parseProblems(lines)

	numRows := len(lines) - 1 // last row is operators
	opRow := lines[len(lines)-1]

	var grandTotal int64
	for _, cols := range problems {
		// Get operator
		opChar := ' '
		for _, c := range cols {
			if c < len(opRow) && opRow[c] != ' ' {
				opChar = rune(opRow[c])
				break
			}
		}

		// Get numbers from rows 0..numRows-1
		var numbers []int64
		for row := 0; row < numRows; row++ {
			line := lines[row]
			// Extract the substring for these columns
			var sb strings.Builder
			for _, c := range cols {
				if c < len(line) {
					sb.WriteByte(line[c])
				} else {
					sb.WriteByte(' ')
				}
			}
			str := strings.TrimSpace(sb.String())
			if str == "" {
				continue
			}
			n, err := strconv.ParseInt(str, 10, 64)
			if err != nil {
				continue
			}
			numbers = append(numbers, n)
		}

		if len(numbers) == 0 {
			continue
		}

		var result int64
		if opChar == '+' {
			for _, n := range numbers {
				result += n
			}
		} else if opChar == '*' {
			result = 1
			for _, n := range numbers {
				result *= n
			}
		}
		grandTotal += result
	}
	return grandTotal
}

// Part 2: Each problem is read right-to-left column by column.
// Each column (within the problem, in the number rows) gives one number,
// where the most significant digit is at the top and least significant at the bottom.
// So each character in a column is one digit of that number.
// Columns of a problem are read right-to-left.
func solvePart2(lines []string) int64 {
	lines = padLines(lines)
	problems := parseProblems(lines)

	numRows := len(lines) - 1 // last row is operators
	opRow := lines[len(lines)-1]

	var grandTotal int64
	for _, cols := range problems {
		// Get operator
		opChar := ' '
		for _, c := range cols {
			if c < len(opRow) && opRow[c] != ' ' {
				opChar = rune(opRow[c])
				break
			}
		}

		// Read columns right-to-left; each column is one number
		var numbers []int64
		for ci := len(cols) - 1; ci >= 0; ci-- {
			col := cols[ci]
			// Collect digit characters from this column, top to bottom (number rows only)
			var sb strings.Builder
			for row := 0; row < numRows; row++ {
				line := lines[row]
				if col < len(line) {
					sb.WriteByte(line[col])
				} else {
					sb.WriteByte(' ')
				}
			}
			str := strings.TrimSpace(sb.String())
			if str == "" {
				continue
			}
			// Each character in this column is a digit; they form the number top-to-bottom
			// TrimSpace removes leading/trailing spaces (from shorter numbers)
			n, err := strconv.ParseInt(str, 10, 64)
			if err != nil {
				// Remove all spaces (in case of internal spaces for some reason)
				str2 := strings.ReplaceAll(str, " ", "")
				if str2 == "" {
					continue
				}
				n2, err2 := strconv.ParseInt(str2, 10, 64)
				if err2 != nil {
					continue
				}
				n = n2
			}
			numbers = append(numbers, n)
		}

		if len(numbers) == 0 {
			continue
		}

		var result int64
		if opChar == '+' {
			for _, n := range numbers {
				result += n
			}
		} else if opChar == '*' {
			result = 1
			for _, n := range numbers {
				result *= n
			}
		}
		grandTotal += result
	}
	return grandTotal
}

func main() {
	lines := readLines("input.txt")

	part := ""
	if len(os.Args) > 1 {
		part = os.Args[1]
	}

	if part == "1" || part == "" {
		fmt.Println("Part 1:", solvePart1(lines))
	}
	if part == "2" || part == "" {
		fmt.Println("Part 2:", solvePart2(lines))
	}
}
